File: ros2_ws/src/behavior_engine/behavior_engine/behaviors/return_anchor.py
"""Return to safe anchor behavior.

Used when battery is critically low or vehicle is damaged.
Returns to a safe anchor point (base station or nearest safe landing zone).
"""

from .base import Behavior

BASE_POSITION = [0.0, 0.0, 0.0]
RETURN_ALT = -20.0


class ReturnToAnchor(Behavior):
    def __init__(self, vehicle_id: str):
        super().__init__(vehicle_id)
        self._anchor = list(BASE_POSITION)
        self._anchor[2] = RETURN_ALT

    def on_enter(self, state: dict):
        # Could choose nearest safe zone — for now always return to base
        self._anchor = [BASE_POSITION[0], BASE_POSITION[1], RETURN_ALT]

    def tick(self, state: dict, fleet: dict, network: dict | None) -> dict:
        return {'position_ned': self._anchor, 'yaw': None}

File: apps/replay-ui/src/sim/autonomy/behaviors/return-anchor.ts
import { Behavior, type VehicleState, type BehaviorResult } from "../behavior";
import type { NetworkResult } from "../../ddil/network-graph";

/** Emergency: return to base station. */
export class ReturnAnchor extends Behavior {
  tick(_state: VehicleState, _fleet: Map<string, VehicleState>, _network: NetworkResult | null): BehaviorResult {
    return { target: [0, 0, -20], yaw: null };
  }
}
